### ISSUE TITLE
[3.0.1] The argument is not initialized or iterable() in phalcon/mvc/view.zep

### ISSUE THREAD

**Issue by ss56806298** [phalcon/cphalcon#12355]
After I updated my phalcon from 2.0.X to 3.0.1,the error occured.And when i turned the version back,it runned normally.I'm not sured where the error happened.

```php
$di = new FactoryDefault();

$di->set('view', function() use ($config) {

    $view = new View();

    return $view;
}, true);
```
### Details

* Phalcon version:3.0.1
* PHP Version: 5.6.25
* Operating System: WINDOWS
* Installation type: DLL
* Server:  Apache
* Other related info (Database, table schema): MYSQL 5.7.14

**Comment by Jurigag** [comment]
Just only this code is causing a problem ? Works fine for me.

**Comment by stamster** [comment]
Just get rid of Windows, it'll work fine on GNU.

**Comment by sergeyklay** [comment]
@ss56806298 Could you please provide detailed error log?

**Comment by ss56806298** [comment]
@stamster 
It's convenient to code on Windows.The error only occured when i updated the phalcon,so i guess the latest version may not suit for Windows.
@sergeyklay 
#0 [internal function]: Phalcon\Mvc\View->_engineRender(Array, 'Login/check', true, true, NULL)
 #1 [internal function]: Phalcon\Mvc\View->render('Login', 'check', Array) 
#2 F:\wamp\www\original_server\index.php(22): Phalcon\Mvc\Application->handle() 
#3 {main}The argument is not initialized or iterable()

**Comment by Jurigag** [comment]
`Login/check'` ? I don't see this in your code. Post whole script to reproduce.

**Comment by SidRoberts** [comment]
It could be that [View::getViewsDirs()](https://github.com/phalcon/cphalcon/blob/v3.0.1/phalcon/mvc/view.zep#L1407) isn't returning an array, [as is required in View::_engineRender()](https://github.com/phalcon/cphalcon/blob/v3.0.1/phalcon/mvc/view.zep#L596). Perhaps if View::setViewsDir() isn't used?

@ss56806298 **if you haven't used View::setViewsDir()**, could you try building Phalcon (using Zephir) from https://github.com/SidRoberts/cphalcon/tree/possible-fix-12355 to see if that fixes the problem?

**Comment by Jurigag** [comment]
If there would be script fully reproduce a problem then it would be good.

**Comment by ss56806298** [comment]
@SidRoberts  @Jurigag 
I alreay solved the problem.Here is the  the structure of my code.

```php
$application = new \Phalcon\Mvc\application($di);
echo $application->handle()->getContent();
```
The variable $di just defined at the top of this page.

LoginController.php
```php
class LoginController extends ControllerBase {
    public function checkAction() {
        try{
            $response = 1;
            $this->sendResponse($response);
        }
    }
}
```

ControllerBase.php
```php
use Phalcon\Mvc\Controller;
```php
use Phalcon\Mvc\Controller;

class ControllerBase extends Controller{
    protected function sendResponse($response) {
        $this->response->setStatusCode(200, 'OK');
        $this->response->setHeader("Content-Type", "application/octet-stream");
        $this->response->setHeader("X-Content-Type-Options", "nosniff");
        $this->response->setContent($content);
        $this->response->send();
    }
}
```

ControllerBase.php after i modified
class ControllerBase extends Controller{
    protected function sendResponse($response) {
        $this->view->disable();
        $this->response->setStatusCode(200, 'OK');
        $this->response->setHeader("Content-Type", "application/octet-stream");
        $this->response->setHeader("X-Content-Type-Options", "nosniff");
        $this->response->setContent($content);
        $this->response->send();
    }
}
```
It seems that the lastest version of phalcon changed at the view.

**Comment by andrew-demb** [comment]
I have encountered the same problem in Debian. As I found, the problem was in `viewsDir`. You can set it before append service `view` in DI or wait for an accept PR @SidRoberts.

**Comment by SidRoberts** [comment]
Returning a response from the controller won't render a view (so you won't have to disable it):

```php
class ControllerBase extends Controller
{
    protected function generateResponse($content)
    {
        $this->response->setStatusCode(200, 'OK');
        $this->response->setHeader("Content-Type", "application/octet-stream");
        $this->response->setHeader("X-Content-Type-Options", "nosniff");
        $this->response->setContent($content);

        return $this->response;
    }
}
```

```php
class LoginController extends ControllerBase
{
    public function checkAction()
    {
        try {
            $response = 1;

            return $this->generateResponse($response);
        } catch (\Exception $e) {
            throw $e;
        }
    }
}
```

**Comment by Jurigag** [comment]
Exactly, just imho always return something from controller action.

**Comment by sergeyklay** [comment]
@SidRoberts It is fixed in `3.0.x`?

**Comment by SidRoberts** [comment]
Yes.
